feat(ios): support multi-scene applications

m1ga opened the pull request:
**Do not merge**
just for building a test SDK

Some info about the original PR and why it was reverted:
Old PR: https://github.com/tidev/titanium-sdk/pull/13941
Issue that caused it to be reverted: https://github.com/tidev/titanium-sdk/issues/13997

Test for the customURLs:
```
const win = Ti.UI.createWindow();

win.addEventListener("open", function () {
  console.log("open", Ti.App.arguments.url)
});

Ti.App.iOS.addEventListener('handleurl', event => {
    console.info("handleurl", event.launchOptions.url);
});

win.open();
```
[setup in this tutorial](https://fromzerotoapp.com/using-custom-url-schemes-to-start-your-app/) and then call it with a custom URL when it's in background and closed. The

m1ga committed:
feat(ios): support multi-scene applications

m1ga committed:
Merge branch 'master' into iOSMultiScene

m1ga committed:
Merge branch 'master' into iOSMultiScene

m1ga committed:
Merge branch 'master' into iOSMultiScene

m1ga committed:
Merge branch 'main' into iOSMultiScene

m1ga committed:
fix registerForPushNotifications callbacks, handleurl

m1ga committed:
Merge branch 'main' into iOSMultiScene

m1ga committed:
remove if

m1ga committed:
remove if

m1ga committed:
Merge branch 'main' into iOSMultiScene

m1ga committed:
fix launch options

m1ga committed:
fix launchoptions

m1ga committed:
fix for lifecycle events

m1ga committed:
arguments update

m1ga committed:
fix handleurl for cold launch

m1ga committed:
fix handleurl for cold launch

m1ga committed:
fix handleurl for cold launch

m1ga committed:
fix handleurl for cold launch

m1ga committed:
Merge branch 'main' into iOSMultiScene

hansemannn commented on iphone/TitaniumKit/TitaniumKit/Sources/API/TiApp.m:
```diff
@@ -421,8 +432,8 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
-  // Boot our kroll-core
-  [self boot];
+  // Skip boot here - the scene delegate (sharedApp) will handle it
+  // This prevents the app from being booted twice
```
**Blocker (root cause).** Boot now happens on the scene-delegate instance, but everything above in this method runs on the *app-delegate* instance and writes per-instance ivars: `localNotification` (404), `remoteNotification` via `generateNotification:` (422), `launchedShortcutItem` (425), and the queued `application:didFinishLaunchingWithOptions:` selector (429). Since `[TiApp app]` == the scene instance, none of these are visible after launch.

For example `NetworkModule.m:255` does `[[TiApp app] remoteNotification]` → reads the scene instance → `nil`, so a cold launch from a remote push loses the payload. Same pattern breaks the launch local notification and the launch shortcut item.

These need to live in shared state (the shared `launchOptions`, or a single shared instance) rather than on whichever instance happens to receive the callback.

hansemannn commented on iphone/TitaniumKit/TitaniumKit/Sources/API/TiApp.m:
```diff
@@ -700,21 +711,33 @@ - (void)invokeSelector:(SEL)selector withArguments:(NSOrderedSet<id> *)arguments
-  NSString *selectorString = NSStringFromSelector(selector);
+  // Only forward if we're the app delegate instance (not the scene delegate / sharedApp)
+  // App delegate has window == nil, scene delegate (sharedApp) has window != nil
+  if (window == nil && sharedApp != nil && sharedApp != self) {
```
Using `window == nil` as the "am I the app delegate?" discriminator is fragile (it's an implementation detail that the app-delegate instance never builds a window). More importantly, at this point during `didFinishLaunchingWithOptions:` `sharedApp` is still `nil`, so this guard is false and the selector gets queued onto the *app-delegate's* `_queuedApplicationSelectors`. `booted:` later flushes the *scene* instance's queue, so anything queued here (e.g. `application:didFinishLaunchingWithOptions:` for modules/Hyperloop) is never delivered. Consider a single owning instance, or forwarding the queued selectors/boot-events to `sharedApp` once it exists.

hansemannn commented on iphone/TitaniumKit/TitaniumKit/Sources/API/TiApp.m:
```diff
@@ -1226,6 +1255,214 @@ - (KrollBridge *)krollBridge
 
+#pragma mark UIWindowSceneDelegate
+
+- (UISceneConfiguration *)application:(UIApplication *)application configurationForConnectingSceneSession:(UISceneSession *)connectingSceneSession options:(UISceneConnectionOptions *)options
```
Memory leak (this file is MRC — `[super dealloc]` / explicit retain-release). This method name doesn't start with `alloc`/`new`/`copy`, so it must return an autoreleased object, but it returns a +1 `UISceneConfiguration` that UIKit won't release → one leak per scene connection. Add `autorelease`:

```objc
return [[[UISceneConfiguration alloc] initWithName:@"Default Configuration" sessionRole:connectingSceneSession.role] autorelease];
```

hansemannn commented on iphone/TitaniumKit/TitaniumKit/Sources/API/TiApp.m:
```diff
@@ -1226,6 +1255,214 @@ - (KrollBridge *)krollBridge
+  [self boot];
+
+  // Handle URL and user activities from scene connection options (iOS 13+ cold launch with URL/intents)
+  NSArray<NSUserActivity *> *userActivities = connectionOptions.userActivities.allObjects;
```
`scene:willConnectToSession:` handles `userActivities` and `URLContexts`, but not `connectionOptions.notificationResponse` (cold launch from a tapped notification) or `connectionOptions.shortcutItem` (cold launch from a Home-screen quick action). These are exactly the launch paths whose data is currently stranded on the app-delegate instance (see the comment in `didFinishLaunchingWithOptions:`). Both should be processed here so notification/shortcut cold-launches reach the booted scene instance.

hansemannn commented on iphone/TitaniumKit/TitaniumKit/Sources/API/TiApp.m:
```diff
@@ -1226,6 +1255,214 @@ - (KrollBridge *)krollBridge
+  }
+
+  // Initialize the root-controller (also sets sharedApp if not already set)
+  [self initController];
```
This runs once **per scene**: `initController` sets `sharedApp = self` and creates a fresh `TiRootViewController`, then `boot` spins up a new `KrollBridge`/JS context. With `UIApplicationSupportsMultipleScenes = true` (set in `Titanium.plist`), opening a second window would overwrite `sharedApp` and boot a second JS runtime — so the app would not actually work as "multi-scene" despite the PR title. If the goal right now is just to adopt the scene lifecycle, keep multiple-scenes disabled; true multi-window needs a per-scene bridge/controller model rather than a single global `sharedApp`.

hansemannn commented on iphone/iphone/Titanium.plist:
```diff
@@ -74,5 +74,22 @@
 	</array>
+	<key>UIApplicationSceneManifest</key>
+	<dict>
+		<key>UIApplicationSupportsMultipleScenes</key>
```
`UIApplicationSupportsMultipleScenes` is `true` here but `false` in the app template `support/iphone/Info.plist:41`. Given the single-global-`sharedApp` design (see `scene:willConnectToSession:`), `true` is unsafe — they should agree, and `false` is the safe value until per-scene state is supported.

hansemannn reviewed (changes requested):
Thorough review of the scene-delegate migration. I understand this is marked **[Do not merge]** / test-SDK only, so this is a design-level review rather than a merge gate — but the issues below look like they explain the original revert (#13997), so worth capturing before this is revived for real.

**Root cause: two `TiApp` instances.** The scene manifest names `TiApp` as the `UISceneDelegateClassName`, and `TiApp` is already the `UIApplicationDelegate`, so iOS creates a *second* `TiApp` instance as the scene delegate. `initController` sets `sharedApp = self` on the **scene** instance, so `[TiApp app]` is the scene instance — but `application:didFinishLaunchingWithOptions:` runs on the **app-delegate** instance. The PR only shares `launchOptions` between the two, so every other piece of launch state captured in `didFinishLaunchingWithOptions:` is stranded on the app delegate and invisible to `[TiApp app]`.

Concrete consequences (inline):
1. `remoteNotification` / `localNotification` set on the app delegate; `NetworkModule.m:255` reads `[[TiApp app] remoteNotification]` (scene instance) → `nil`. Launch push payload / local notification are lost.
2. `launchedShortcutItem` set on the app delegate; `booted:` (scene instance) checks its own nil ivar, and `scene:willConnectToSession:` never reads `connectionOptions.shortcutItem` → quick-action cold launch does nothing.
3. Selectors queued during `didFinishLaunchingWithOptions:` land on the app delegate's queue; `booted:` flushes the scene's queue → module/Hyperloop `application:didFinishLaunchingWithOptions:` forwarding is dropped.

**Other issues:** a `UISceneConfiguration` leak under MRC, and the design doesn't actually support multiple scenes (each scene re-runs `initController`+`boot`, clobbering `sharedApp` and booting a second JS context) despite the title — see inline comments.

**Suggested direction:** capture launch notification/shortcut state into the shared `launchOptions` (or a single shared instance) instead of per-instance ivars, and handle `connectionOptions.notificationResponse` + `.shortcutItem` in `scene:willConnectToSession:`. Either fully commit to one shared instance or make the forwarding cover *all* launch state, not just `launchOptions`.

Happy to re-review once the dual-instance state-splitting is resolved.

hansemannn committed:
Merge branch 'main' of https://github.com/tidev/titanium-sdk into iOSMultiScene

hansemannn committed:
fix(ios): resolve scene-delegate state splitting in multi-scene support

Addresses review feedback on the scene-lifecycle migration (#14195):

- Route all scene callbacks to the single owning (app-delegate) instance so
  launch state captured in didFinishLaunchingWithOptions: stays visible via
  [TiApp app]. Removes the fragile `window == nil` discriminator.
- boot/booted: now run on the same instance that queued launch selectors, so
  module/Hyperloop application:didFinishLaunchingWithOptions: forwarding is
  delivered instead of stranded on the orphan instance.
- Handle connectionOptions.notificationResponse and .shortcutItem on cold
  launch, plus windowScene:performActionForShortcutItem: for warm quick
  actions, so notification/shortcut launches reach the booted instance.
- Fix UISceneConfiguration leak under MRC (autorelease).
- Set UIApplicationSupportsMultipleScenes to false to match the app template
  and the current single-shared-instance design.

hansemannn committed:
chore: rename owningInstance => owningApp

hansemannn committed:
fix: revert renaming

m1ga committed:
Merge branch 'main' into iOSMultiScene

hansemannn reviewed (approved):
Let's go with this initial version in 13.3.0 and extend it further as part of the next updates. Thanks everyone!
